src/Npgsql/NpgsqlTypes/NpgsqlTsQuery.cs

```diff
@@ -227,42 +299,95 @@ public static NpgsqlTsQuery Parse(string value)
-                }
-                else if (opStack.Peek() == '!')
-                {
+                    var andRight = valStack.Pop();
+                    var andLeft = valStack.Pop();
+                    valStack.Push(new NpgsqlTsQueryAnd(andLeft, andRight));
+                    opStack.Pop();
+                    break;
+
+                case '!':
                     if (valStack.Count == 0)
                         throw new FormatException("Syntax error in tsquery");
                     valStack.Push(new NpgsqlTsQueryNot(valStack.Pop()));
+                    opStack.Pop();
+                    break;
+
+                case '<':
+                    if (valStack.Count < 2)
+                        throw new FormatException("Syntax error in tsquery");
+                    var followedByRight = valStack.Pop();
+                    var followedByLeft = valStack.Pop();
+                    valStack.Push(
+                        new NpgsqlTsQueryFollowedBy(
+                            followedByLeft,
+                            tsOp.FollowedByDistance,
+                            followedByRight));
+                    opStack.Pop();
+                    break;
+
+                default:
+                    processTightBindingOperator = false;
+                    break;
                 }
-                opStack.Pop();
             }
             expectingBinOp = true;
             goto NextToken;
 
             Finish:
             while (opStack.Count > 0)
             {
-                if (valStack.Count < 2 || (opStack.Peek() != '|' && opStack.Peek() != '&'))
+                if (valStack.Count < 2)
                     throw new FormatException("Syntax error in tsquery");
+
                 var right = valStack.Pop();
                 var left = valStack.Pop();
-                valStack.Push(opStack.Pop() == '&' ? (NpgsqlTsQuery)new NpgsqlTsQueryAnd(left, right) : new NpgsqlTsQueryOr(left, right));
+
+                NpgsqlTsQuery query;
+                var tsOp = opStack.Pop();
+                switch (tsOp)
+                {
+                case '&':
+                    query = new NpgsqlTsQueryAnd(left, right);
+                    break;
+
+                case '|':
+                    query = new NpgsqlTsQueryOr(left, right);
+                    break;
+
+                case '<':
+                    query = new NpgsqlTsQueryFollowedBy(left, tsOp.FollowedByDistance, right);
+                    break;
+
+                default:
+                    throw new FormatException("Syntax error in tsquery");
+                }
+
+                valStack.Push(query);
             }
             if (valStack.Count != 1)
                 throw new FormatException("Syntax error in tsquery");
             return valStack.Pop();
         }
     }
 
+    struct NpgsqlTsQueryOperator
+    {
+        public char Char { get; set; }
+        public int FollowedByDistance { get; set; }
```

Comment:
It's an internal struct. You can use public fields instead of properties. And probably it would be better to make them readonly.

Reply:
Is it OK to use C# 7.2 features in Npgsql ? @YohDeadfall